ament_cmake_gen_version_h/cmake/ament_cmake_gen_version_h.cmake

```diff
@@ -0,0 +1,84 @@
+# the destination `version.h` file and installs that generated file into `DESTINATION include`.
+# The generated file is being (re-)created if:
+# - the file does not exist
+# - the file does exists but contains a version that differs from the version in `package.xml` file
+################################################################################
+function(ament_cmake_gen_version_h)
+  if (NOT ${ARGC} EQUAL 0)
+    message(FATAL_ERROR "ament_cmake_gen_version_h does not accept parameters but ${ARGC} parameters provided")
+  endif()
+
+  # Find myself so I can use my own cmake instalation folder `ament_cmake_gen_version_h_DIR`
+  # I can't rely on a previous find_package call because it could be a direct call
+  find_package(ament_cmake_gen_version_h REQUIRED)
+  set(TEMPLATE_FILE "${ament_cmake_gen_version_h_DIR}/version.h.in")
+  if(NOT EXISTS "${TEMPLATE_FILE}")
+    message(FATAL_ERROR "Can't find ${TEMPLATE_FILE}. Reinstall ament_cmake_gen_version_h package.")
+  endif()
+
+  include_directories(${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_gen_version_h/include)
+  set(TMP_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_gen_version_h/include/${PROJECT_NAME})
+
+  set(VERSION_FILE_NAME ${TMP_INCLUDE_DIR}/version.h)
+  set(NEED_TO_CREATE_VERSION_FILE TRUE)
+
+  # retrieve version information from <package>.xml file
+  # call ament_package_xml() if it has not been called before
+  if(NOT _AMENT_PACKAGE_NAME)
+    ament_package_xml()
+  endif()
+
+  string(TOUPPER ${PROJECT_NAME} PROJECT_NAME_UPPER)
+  set(VERSION_STR ${${PROJECT_NAME}_VERSION})
+
+  # parse version information from the version string
+  string(REGEX MATCH "([0-9]+)\.([0-9]+)\.([0-9]+)" "" dummy ${VERSION_STR})
+  set(VERSION_MAJOR ${CMAKE_MATCH_1})
+  set(VERSION_MINOR ${CMAKE_MATCH_2})
+  set(VERSION_PATCH ${CMAKE_MATCH_3})
+
+  # Check if the version file exist
+  if(EXISTS "${VERSION_FILE_NAME}")
+    # The file exists
+    # Check if it contains the same version
+    set(LINE_PATTERN "#define[ \t]+${PROJECT_NAME_UPPER}_VERSION_STR")
+    file(STRINGS ${VERSION_FILE_NAME} VERSION_FILE_STRINGS REGEX ${LINE_PATTERN})
+    set(VERSION_PATTERN "^#define[ \t]+${PROJECT_NAME_UPPER}_VERSION_STR[ \t]+\"([0-9]+\.[0-9]+\.[0-9]+)\"")
+    string(REGEX MATCH ${VERSION_PATTERN} dummy ${VERSION_FILE_STRINGS})
+    if("${CMAKE_MATCH_1}" STREQUAL "${VERSION_STR}")
+      message(STATUS "File and project versions match: \"${CMAKE_MATCH_1}\" == \"${VERSION_STR}\"")
+      set(NEED_TO_CREATE_VERSION_FILE FALSE)
+    endif()
+  endif()
+
+  if(${NEED_TO_CREATE_VERSION_FILE})
+    message(STATUS "Create new version file for version ${${PROJECT_NAME}_VERSION}")
+    file(MAKE_DIRECTORY ${TMP_INCLUDE_DIR})
+    # create the version.h file
+    configure_file(${TEMPLATE_FILE} ${VERSION_FILE_NAME})
+  else()
+    message(STATUS "Skip version file creation")
+  endif()
+
+  install(
+    DIRECTORY ${TMP_INCLUDE_DIR}
+    DESTINATION include)
```

Comment:
@serge-nikulin I'm somewhat inclined towards generating the file and letting the caller decide if and where to install it (as if it was an OUTPUT of `add_custom_command`). 

I'd be OK with optionally delegating the installation though.

Reply:
@hidmic, I decided to parametrize pretty much everything. How do the following optional parameters look to you?
If you are OK, I'll proceed with them and also create some unit tests for different scenarios.

```cmake
# :param DO_NOT_INSTALL: whether to autmatically install the generated version file into DESTINATION include
# :type DO_NOT_INSTALL: BOOL
# :default value DO_NOT_INSTALL: FALSE

# :param PRJ_NAME: project name to use
# :type PRJ_NAME: string
# :default value PRJ_NAME: ${PROJECT_NAME}

# :param INCLUDE_DIR: path to the include folder where the file will be generated
#     ${INCLUDE_DIR} folder will be added to the include paths
#     the file will be placed into ${INCLUDE_DIR}/${PRJ_NAME} folder according to ROS2 standard
# :type INCLUDE_DIR: string
# :default value INCLUDE_DIR: ${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_gen_version_h/include

# :param VERSION_FILE_NAME: file name of the generated header file
# :type VERSION_FILE_NAME: string
# :default value VERSION_FILE_NAME: version.h
```

Reply:
@serge-nikulin may I suggest a `NO_INSTALL` option instead of a `DO_NOT_INSTALL` one value argument? Also, what use case do you have in mind for `PRJ_NAME`?

Reply:
Agree with both points. Please review again.